src/AstGen.zig

```diff
@@ -1718,6 +1718,61 @@ fn structInitExpr(
         }
     }
 
+    // Temporary Arena to simplify freeing the HashMaps
+    var dup_arena = std.heap.ArenaAllocator.init(astgen.arena);
+    const dup_alloc = dup_arena.allocator();
+
+    // Key: The first found variable.
+    // Value: A list of following duplicate variable names.
+    var duplicate_names = std.AutoHashMap(Ast.TokenIndex, std.ArrayList(Ast.TokenIndex)).init(dup_alloc);
```

Comment:
The key here is not "the first found variable" (nit: "variable" is never a relevant term in this code, they're fields!), nor is it an `Ast.TokenIndex`. It is a plain `u32` which is the string index of the field name. (It's understandable that you messed up the types here, we really need to make these type aliases safer)

Also, using an arraylist for the value here is alright, but please make it an `ArrayListUnmanaged`. This avoids the need to duplicate the allocator N times in memory.

To optimize memory usage, you could change the key type to be something like a `const DupeNameRecord = struct { first: Ast.TokenIndex, duplicates: ArrayListUnmanaged(Ast.TokenIndex) }`. This structure means that in the success path, you don't need to allocate any memory to `duplicates`, so freeing `duplicate_names` from the arena at the end could work.